File: uber/uber_algo/LC236LowestCommonAncestorofaBinaryTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        if not root:
            return None
        queue = [root]
        dictionary = {root: None}
        while p not in dictionary or q not in dictionary:
            element = queue.pop(0)
            if element.left:
                dictionary[element.left] = element
                queue.append(element.left)
            if element.right:
                dictionary[element.right] = element
                queue.append(element.right)
        temp = set()
        while p:
            temp.add(p)
            p = dictionary[p]
        while q not in temp:
            q = dictionary[q]
        return q

File: apple/LC236LowestCommonAncestorOfABinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        Map<TreeNode, TreeNode> hashmap = new HashMap<>();
        hashmap.put(root, null);
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        while (!hashmap.containsKey(p) || !hashmap.containsKey(q)) {
            TreeNode node = queue.poll();
            if (node.left != null) {
                hashmap.put(node.left, node);
                queue.offer(node.left);
            }
            if (node.right != null) {
                hashmap.put(node.right, node);
                queue.offer(node.right);
            }
        }
        Set<TreeNode> answer = new HashSet<>();
        while (p != null) {
            answer.add(p);
            p = hashmap.get(p);
        }
        while (!answer.contains(q)){
            q = hashmap.get(q);
        }
        return q;
    }
}
